local/toolchest/nmake/src/lib/libx/hash/hash.3
.TH HASH 3
.SH NAME
hash \- hash table library
.SH SYNOPSIS
\f5#include <hash.h>\fP
.SH DESCRIPTION
The
.I hash
library manipulates generalized hash tables.
Many different hash tables can be instantiated within a single program.
.PP
A hash table provides an association between a
.I key
and its
.IR value .
A
.I key
is a sequence of 
.B char
elements
and the
.I value
is a user supplied pointer to the value.
Each hash table has
.I size
slots,
and each slot is a pointer to the head of a forward linked
.I bucket
chain.
.PP
Hashing occurs as follows:
a
.I "hash function"
takes 
.I key
and 
.I size
as arguments and produces an index of a
.I slot
in the hash table.
The 
.I "bucket chain"
corresponding to the slot
is then sequentially searched until a
match is found for
.I key
using the user supplied comparison routine.
.PP
Three basic types are defined in the header file
.BR hash.h :
.TP
.B GENERIC
This is the type of the
.I value
associated with each
.I key
in the hash.
It is currently
.B char*
but will eventually become
.BR void* .
.TP
.B TABLE
A variable of this type points to a hash table structure.
The public structure elements are:
.RS
.TP
.B "unsigned int flags"
The hash table flags.
The flags are initialized by the
.I mkhash
call and may be modified after the table has been created.
.TP
.B "BUCKET lastbucket"
A pointer to the most recent hash bucket used by either
.I gethash
or
.IR puthash .
.TP
.B "int chains"
The number of bucket chains.
This number can be used to estimate the effectiveness of the hashing function.
.PP
If the
.I hash
library was compiled with the
.B HASH_INFO
compile option set then the following are also available:
.TP
.B "int accesses"
The number of 
.I gethash
and
.I puthash
calls placed on the current table.
.TP
.B "int buckets"
The number of hash buckets.
.TP
.B "int collisions"
The number of failed key comparisons.
.RE
.TP
.B BUCKET
A variable of this type points to a 
.IR "hash bucket" .
The public structure elements are:
.RS
.TP
.B "char *name"
A pointer to the hash bucket key.
.TP
.B "GENERIC value"
The value associated with the key.
.RE
.PP
The hash routines are:
.TP
\f5TABLE mkhash(unsigned int size, int flags, unsigned int (*hash)(),\fP
.PD 0
.TP
	\f5int (*compare)(), GENERIC error, char* name)\fP
.I mkhash
makes a new hash table and returns a pointer to the table.
.I malloc (3)
is used to allocate space for the table.
NULL is returned if the table cannot be created.
The arguments are:
.RS
.TP
.B size
The number of slots in the table.
This is the maximum number of bucket chains for the table.
.TP
.B flags
Specifies certain hash table properties.
The flags may be 
.IR or ed
together.
.B HASH_ALLOCATE
specifies that the key names are null-terminated strings
and that new keys entered by
.I puthash
are to be copied to data areas obtained from
.IR malloc (3).
.B HASH_DATA
specifies that the key names are binary data and are
not to be treated as null-terminated strings.
.B HASH_POWER2
specifies that the log base 2 of the table size is to be
passed as the second argument to the hashing function.
.B HASH_REMOVE
specifies that 
.I puthash
is to delete table entries with the value
.BR error .
.TP
.B hash
A pointer to the hashing function for the table.
If
.I hash
is NULL then a default null-terminated string hash algorithm is used.
.TP
.B compare
A pointer to the hash key comparison function.
The return value of
.I compare
should be the same as for
.IR strcmp (3).
The first argument is the 
.I key
form the current hash bucket on the 
.I "collision chain"
and the second argument is the user supplied 
.IR key .
If
.I compare
is NULL then a default null-terminated string comparison algorithm is used.
.TP
.B error
The value returned by
.I gethash
for elements not found in the table.
.TP
.B name
The hash table name that is used by
.IR dumphash .
.RE
.TP
\f5void rmhash(TABLE tab)\fP
.I rmhash
removes a hash table created by
.IR mkhash .
The caller is responsible for removing the
.I values
associated with each
.IR key .
The arguments are:
.RS
.TP
.B tab
The hash table pointer returned by
.IR mkhash .
.RE
.TP
\f5TABLE rehash(TABLE tab, unsigned int size)\fP
.I rehash
changes the size of hash table
.I tab
created by
.IR mkhash .
A pointer to the re-hashed table is returned.
The old table is removed and a pointer to the new table is returned.
NULL is returned if the new table could not be generated.
The old table is removed only for successful return values.
The arguments are:
.RS
.TP
.B tab
The hash table pointer of the old table.
.TP
.B size
The new hash table size.
.RE
.TP
\f5void exhash(TABLE tab, int (*examiner)())\fP
.I exhash
examines (walks through) a hash table and applies a user supplied
function to each entry in the table.
The arguments are:
.RS
.TP
.B tab
The hash table pointer.
.TP
.B examiner
The function applied to each entry in the table.
The function must return 0 on success, otherwise 
.I exhash
returns before all the table elements have been examined.
.I examiner
is declared as
.nf
    \f5int examiner(char *name, GENERIC value)\fP
.fi
where 
.I name
is the key name of the current hash table entry and
.I value
is the value associated with the key.
.RE
.TP
\f5GENERIC gethash(TABLE tab, char* key)\fP
.I gethash
returns the value associated with
.I key
in table
.IR tab .
If no value is associated with 
.I key
then the argument
.I error
from the corresponding 
.I mkhash
call is returned.
The arguments are:
.RS
.TP
.B tab
The hash table pointer.
.TP
.B key
The key name.
.RE
.TP
\f5char* puthash(TABLE tab, char* key, GENERIC value)\fP
.I puthash
associates
the pointer
.I value
with
.I key
in the hash table
.IR tab .
If the 
.B HASH_ALLOCATE
flag is on and the
.B HASH_DATA
flag is off then the null-terminated string
.I key
is copied to a non-volatile area maintained by the
.I hash
library.
A pointer to the possibly copied 
.I key
is returned.
The arguments are:
.RS
.TP
.B tab
The hash table pointer.
.TP
.B key
The key name.
If
.I key
is NULL and the
.B HASH_ALLOCATE
flag is set then
.I key
is assumed to be the key from the most recent
.I puthash
call.
This allows 
.I gethash-puthash
pairs to be optimized by retaining information from the
.I gethash
call.
.TP
.B value
A pointer to the key value.
If the
.B HASH_REMOVE
flag is set and
.I value
is equal to the value of the argument
.I error
to the 
.I mkhash 
call then
.I key 
is removed from the hash table.
.RE
.TP
\f5void dumphash(TABLE tab, int flags)\fP
If the
.I hash
library was compiled with
.B HASH_INFO
defined then
.I dumphash
dumps the hash table accounting on the standard error, otherwise
a warning message is output.
The arguments are:
.RS
.TP
.B tab
The hash table pointer.
If
.I tab
is
NULL then information on all tables generated by
.I mkhash
is dumped.
.TP
.B flags
If 
.I flags
is 0 then only the
.IR accesses ,
.I buckets
and
.I collisions
accounting is listed, otherwise all
.I key-value
pairs are listed for
.IR tab .
.RE
.TP
\f5unsigned int chash(char* key, unsigned int size)\fP
.I chash
hashes the character string
.I key
and returns an integer between 0 and 
.IR size \-1
inclusively.
The arguments are:
.RS
.TP
.B key
The character string to be hashed.
.TP
.B size
The size of the hash table.
.RE
.TP
\f5unsigned int chash2(char* key, unsigned int log)\fP
.I chash2
is similar to 
.I chash
except that it is optimized for hash tables with size a power of 2.
The arguments are:
.RS
.TP
.B key
The character string to be hashed.
.TP
.B log
The log base 2 of the hash table size.
.RE
.SH "SEE ALSO"
hsearch(3)


local/toolchest/nmake/src/lib/libx/hash/hash.c
/*

 *      Copyright (c) 1984, 1985 AT&T
 *      All Rights Reserved

 *      THIS IS UNPUBLISHED PROPRIETARY SOURCE 
 *      CODE OF AT&T.
 *      The copyright notice above does not 
 *      evidence any actual or intended
 *      publication of such source code.

 */
/*
 * hash routines
 *
 * G. S. Fowler
 * AT&T Bell Laboratories
 *
 * HASH_INFO enables hash monitoring
 */

#include <stdio.h>
#include <hash.h>

static char	stamp[] = "@(#)hash (AT&T Bell Laboratories) 08/30/85a";

#define HASH(t,n,a,v)	if((t)->hash)v=(*t->hash)(n,a);\
			else{register char *s=n;register int i=0;\
			while(*s)i=(i<<1)+*s++;if((i%=(t)->size)<0)i+=(t)->size;v=(unsigned int)i;}

extern char	*malloc();
extern char	*strcpy();

extern int	strlen();

#if HASH_INFO

static void	hashinfo();
static void	hashrm();

static TABLE	tables;

#else

#define hashrm(t)	free(t)

#endif

/*
 * create a new hash table
 */

TABLE
mkhash(size, flags, hash, compare, error, name)
register unsigned int	size;
unsigned int		flags;
unsigned int		(*hash)();
int			(*compare)();
GENERIC			error;
char			*name;
{
	register TABLE	tab;
	register BUCKET	*b;
	unsigned int	i;

	if (!(tab = (TABLE)malloc(sizeof(struct hashtable)))) return((TABLE)0);
	if (!(b = (BUCKET*)malloc(sizeof(BUCKET) * size)))
	{
		free(tab);
		return((TABLE)0);
	}
	tab->table = b;
	tab->size = size;
	if (flags & HASH_POWER2) for (i = size, tab->arg = 1; i >>= 1; tab->arg++);
	else tab->arg = size;
	tab->flags = flags;
	tab->hash = hash;
	tab->compare = compare;
	tab->error = error;
	while (size--) *b++ = (BUCKET)0;
	tab->chains = 0;
#if HASH_INFO
	tab->name = name;
	tab->accesses = 0;
	tab->buckets = 0;
	tab->collisions = 0;
	tab->next = tables;
	tables = tab;
#endif
	return(tab);
}

/*
 * remove a hash table
 */

void
rmhash(tab)
register TABLE	tab;
{
	register unsigned int	i;
	register BUCKET		*b, k;

	b = tab->table;
	i = tab->size;
	while (i--)
		for (k = *b++; k; k = k->next)
		{
			if (tab->flags & HASH_ALLOCATE) free(k->name);
			free(k);
		}
	free(tab->table);
	hashrm(tab);
}

/*
 * rehash an old table
 */

TABLE
rehash(tab, size)
TABLE		tab;
unsigned int	size;
{
	unsigned int	i, index;
	TABLE		new;
	register BUCKET	*tab_b, tab_k, *new_b, new_k;

	if (size > 0 && tab->size != size)
	{
#if HASH_INFO
		if (!(new = mkhash(size, tab->flags, tab->hash, tab->compare, tab->error, tab->name))) return((TABLE)0);
		new->buckets = tab->buckets;
#else
		if (!(new = mkhash(size, tab->flags, tab->hash, tab->compare, tab->error, (char*)0))) return((TABLE)0);
#endif
		tab_b = tab->table;
		new_b = new->table;
		i = tab->size;
		while (i--)
		{
			tab_k = *tab_b++;
			while (tab_k)
			{
				new_k = tab_k;
				tab_k = tab_k->next;
				HASH(tab, new_k->name, new->arg, index);
				if (!(new_k->next = new_b[index])) new->chains++;
				new_b[index] = new_k;
			}
		}
		hashrm(tab);
		return(new);
	}
	return(tab);
}

/*
 * examine (walk through) a hash table
 */

void
exhash(tab, examiner)
TABLE	tab;
int	(*examiner)();
{
	register unsigned int	i;
	register BUCKET		*b, k;

	b = tab->table;
	i = tab->size;
	tab->flags |= HASH_EXAMINE;
	while (i--)
		for (k = *b++; k; k = k->next)
			if (k->value != tab->error && (*examiner)(k->name, k->value) < 0)
			{
				tab->flags &= ~HASH_EXAMINE;
				return;
			}
	tab->flags &= ~HASH_EXAMINE;
}

/*
 * get value for name
 */

GENERIC
gethash(tab, name)
register TABLE	tab;
char		*name;
{
	register BUCKET	b, p;

#if HASH_INFO
	tab->accesses++;
#endif
	tab->lastname = name;
	tab->lastbucket = (BUCKET)0;
	HASH(tab, name, tab->arg, tab->lastindex);
	for (p = (BUCKET)0, b = tab->table[tab->lastindex]; b; p = b, b = b->next)
	{
		if (tab->compare)
		{
			if (!(*tab->compare)(b->name, name)) goto found;
#if HASH_INFO
			tab->collisions++;
#endif
			continue;
		}
		else
		{
			register char	*s1 = b->name, *s2 = name;

			/*
			 * default string comparison
			 */

			while (*s1 == *s2++)
				if (!*s1++) goto found;
#if HASH_INFO
			tab->collisions++;
#endif
			continue;
		}

		/*
		 * found a match
		 */

	found:
		tab->lastbucket = b;
		if (p && !(tab->flags & HASH_EXAMINE))
		{
			/*
			 * move this bucket to the front of the list
			 * this places the popular names near the front
			 */

			p->next = b->next;
			b->next = tab->table[tab->lastindex];
			tab->table[tab->lastindex] = b;
		}
		return(b->value);
	}
	return(tab->error);
}

/*
 * put name=value
 *
 * the code is optimized for the sequence
 *
 * 	if (gethash(tab, name) == tab->error) puthash(tab, (char*)0, value);
 *
 * which is equivalent to but more efficient than
 *
 * 	if (gethash(tab, name) == tab->error) puthash(tab, name, value);
 */

char*
puthash(tab, name, value)
register TABLE	tab;
char		*name;
GENERIC		value;
{
	unsigned int	index;
	register BUCKET	b;

	if (name || (tab->flags & HASH_DATA)) (void)gethash(tab, name);
	else
	{
#if HASH_INFO
		tab->accesses++;
#endif
		name = tab->lastname;
	}
	if (b = tab->lastbucket)
	{
#if HASH_INFO
		if (value == tab->error) tab->buckets--;
		else if (b->value == tab->error) tab->buckets++;
#endif
		if ((b->value = value) == tab->error &&
		    (tab->flags & (HASH_EXAMINE | HASH_REMOVE)) == HASH_REMOVE)
		{
			if (tab->flags & HASH_ALLOCATE) free(b->name);
			tab->table[tab->lastindex] = b->next;
			free(b);
			return((char*)0);
		}
		return(b->name);
	}
	else if (value == tab->error) return((char*)0);
	else index = tab->lastindex;
	if (!(b = (BUCKET)malloc(sizeof(struct hashbucket)))) return((char*)0);
	if (!(b->next = tab->table[index])) tab->chains++;
	tab->table[index] = tab->lastbucket = b;
	if (tab->flags & HASH_ALLOCATE)
	{
		if (!(b->name = malloc(strlen(name)+1)))
		{
			tab->table[index] = b->next;
			free(b);
			return((char*)0);
		}
		(void)strcpy(b->name, name);
	}
	else b->name = name;
#if HASH_INFO
	tab->buckets++;
#endif
	b->value = value;
	return(b->name);
}

/*
 * dump hash table accounting info
 * if p is 0 then dump all tables
 */

void
dumphash(tab)
register TABLE	tab;
{
#if HASH_INFO
	fprintf(stderr, "\nhash table information:\n\n");
	if (tab) hashinfo(tab);
	else for (tab = tables; tab; tab = tab->next) hashinfo(tab);
	fflush(stderr);
#else
	static int	first = 1;

	if (first)
	{
		first = 0;
		fprintf(stderr, "hash: compile hash routines with -DHASH_INFO to get detailed hash table accounting\n");
		fflush(stderr);
	}
#endif
}

#if HASH_INFO

/*
 * dump info on a single table
 */

static void
hashinfo(tab)
register TABLE	tab;
{
	fprintf(stderr, "\tname:        %s\n", tab->name);
	fprintf(stderr, "\taddress:     0x%06lx\n", (unsigned long)tab);
	fprintf(stderr, "\tsize:        %d\n", tab->size);
	fprintf(stderr, "\tflags:       ");
	if (tab->flags & HASH_ALLOCATE) fprintf(stderr, "[allocate] ");
	if (tab->flags & HASH_DATA) fprintf(stderr, "[data] ");
	if (tab->flags & HASH_POWER2) fprintf(stderr, "[power/2] ");
	if (tab->flags & HASH_REMOVE) fprintf(stderr, "[remove] ");
	fprintf(stderr, "\n");
	fprintf(stderr, "\taccesses:    %d\n", tab->accesses);
	fprintf(stderr, "\tbuckets:     %d\n", tab->buckets);
	fprintf(stderr, "\tchains:      %d\n", tab->chains);
	fprintf(stderr, "\tcollisions:  %d\n", tab->collisions);
	fprintf(stderr, "\n");
	fflush(stderr);
}

/*
 * remove a table from the accounting list
 */

static void
hashrm(tab)
register TABLE	tab;
{
	register TABLE	q;

	if (tab == tables) tables = tab->next;
	else for (q = tables; q; q = q->next)
		if (tab == q->next)
		{
			q->next = tab->next;
			break;
		}
	free(tab);
}

#endif
